syntaxes/modsecurity.tmLanguage.json

```diff
@@ -0,0 +1,736 @@
+    {
+      "captures": {
+        "2": {
+          "name": "keyword.operator.modsecurity.action"
+        },
+        "4": {
+          "name": "constant.numeric.modsecurity.action_parameter"
+        },
+        "5": {
+          "name": "constant.numeric.modsecurity.action_parameter"
+        },
+        "6": {
+          "name": "string.quoted.modsecurity.action_parameter"
+        },
+        "7": {
+          "name": "string.quoted.modsecurity.action_parameter"
+        }
+      },
+      "comment": "Rule action - with parameter",
+      "match": "(?i)(^|,|\\\"|\\s)(accuracy|append|deprecatevar|exec|expirevar|id|initcol|logdata|maturity|msg|pause|prepend|rev|sanitiseArg|sanitiseMatched|sanitiseMatchedBytes|sanitiseRequestHeader|sanitiseResponseHeader|setuid|setrsc|setsid|setenv|setvar|skip|skipAfter|status|tag|ver|xmlns):('(\\d+)'|(\\d+)|('[\\w\\s:\\.\\-_/\\(\\),]+')|([\\w\\s:\\.\\-_/\\(\\)]+))",
+      "name": "complex.modsecurity.action.with_parameter"
+    },
+    {
+      "captures": {
+        "2": {
+          "name": "keyword.operator.modsecurity.action"
+        }
+      },
+      "comment": "Rule action - with parameter not yet entered, or unmatched (e.g. setvar with macros)",
+      "match": "(?i)(^|,|\\\"|\\s)(accuracy|append|deprecatevar|exec|expirevar|id|initcol|logdata|maturity|msg|pause|phase|prepend|ev|sanitiseArg|sanitiseMatched|sanitiseMatchedBytes|sanitiseRequestHeader|sanitiseResponseHeader|severity|setuid|setrsc|setsid|setenv|setvar|skip|skipAfter|status|tag|ver|xmlns):",
+      "name": "complex.modsecurity.action.with_parameter.other"
+    },
+    {
+      "captures": {
+        "2": {
+          "name": "entity.name.function.modsecurity.action.disruptive_pass"
+        }
+      },
+      "comment": "Rule action - with parameter - disruptive actions",
+      "match": "(?i)(^|,|\\\"|\\s)(proxy|redirect):",
+      "name": "complex.modsecurity.action.with_parameter.disruptive"
+    },
+    {
+      "captures": {
+        "2": {
+          "name": "entity.name.function.modsecurity.action.disruptive_pass"
+        }
+      },
+      "comment": "Rule action - without parameters - disruptive actions, pass, capture, and chain",
+      "match": "(?i)(^|,|\\\"|\\s)(allow|block|deny|drop|pass|chain)\\b",
+      "name": "complex.modsecurity.action.without_parameter.disruptive_pass"
+    },
+    {
+      "captures": {
+        "2": {
+          "name": "keyword.operator.modsecurity.action"
+        }
+      },
+      "comment": "Rule action - without parameters",
+      "match": "(?i)(^|,|\\\"|\\s)(auditlog|capture|log|multiMatch|noauditlog|nolog)\\b",
+      "name": "complex.modsecurity.action.without_parameter"
+    },
+    {
+      "captures": {
+        "1": {
+          "name": "keyword.control.modsecurity"
+        },
+        "2": {
+          "name": "string.regexp.modsecurity"
+        }
+      },
+      "comment": "Regexp operator and operand, e.g. \"@rx foo\"",
+      "match": "(?i)\"(@rx)\\s+((?:\\\"|[^\"])*)",
```

Comment:
This RegEx `(?:\"|[^"])` means `"` or non-`"`, which effectively makes it all characters, was that what you were going for?

The regex is tricky for this one.. But this SO answer helps: https://stackoverflow.com/a/249937/2037090

```
(?i)"(@rx)\s+((?:[^"\\]|\\.)*)"
```

Reply:
Looks working when I added an escaped double quote in a CRS rule.

![image](https://user-images.githubusercontent.com/9884746/158855467-3a3d105b-b7ef-4fb6-bb74-6c7853ed4610.png)

Reply:
Btw.. It seems like rules starting with `!@rx` are not handled and are treated as string literals.

![image](https://user-images.githubusercontent.com/9884746/158856776-f252ce49-d491-4b11-b7dd-e96c30069dc7.png)

I really don't remember any ModSecurity syntax any more, but it seems like it should mean "not matching regex" right? I guess we could handle those as well, we could create a separate issue for that.

Reply:
Yes, let's do it in a different PR. I can confirm that this is the behavior, but is not expected.

Reply:
Yeah, let's create a new issue and PR with that.

Reply:
Sorry for the confusion. I meant that the regex did not work, you have to use

```
(?i)"(@rx)\s+((?:[^"\\]|\\.)*)"
```

Reply:
Hmm, still not right, you need to escape the `\` as well.

After escaping it should be 

```
"match": "(?i)\"(@rx)\\s+((?:[^\"\\\\]|\\\\.)*)\"",
```

Reply:
🤦 Sorry! :) Looks like it is good now.